Classes/Common/Helper.php

```diff
@@ -1068,5 +1070,39 @@ public static function getLanguageService(): LanguageService
         return $GLOBALS['LANG'];
     }
 
+    /**
+     * Make classname configuration from `Classes.php` available in contexts
+     * where it normally isn't, and where the classical way via TypoScript won't
+     * work either.
+     *
+     * @deprecated Remove once we drop support for TYPO3@9
+     *
+     * @access public
+     */
+    public static function polyfillExtbaseClassesForTYPO3v9()
+    {
```

Comment:
The table mapping is done in TYPO3 9 in `ext_typoscript_setup.txt`. In TYPO3 10 it is done in `Configuration/Extbase/Persistence/Classes.php`. Both should coexist without problems - I thought.

Reply:
They probably can coexist, but in v9 neither works for use in expression language functions (`DocumentTypeFunctionProvider`). In testing I got an error that the table `tx_dlf_domain_model_documents` does not exist, i.e., it took the default tablename instead of reading the configuration.

I don't know enough about TYPO3 to tell exactly why, but it's probably related to expression language working within TypoScript as well. I guess that `ext_typoscript_setup.txt` is loaded relatively early, but perhaps it hasn't yet been processed by Extbase when the `getDocumentType()` is evaluated within another TypoScript.

Reply:
I remember this behaviour. And I think, there was in the meantime in `DocumentTypeFunctionProvider` a patch for the mapping which I might have removed again. I'll will check again.

Reply:
In 5e6110fb you added some code to `Document::getInstance()` for that (and removed it later). I basically generalized this to take all classnames from `Classes.php`.

Reply:
Exactly. That's what I meant. Ok, I understand your code now.

We could drop all this, if we rename the tables to extbase default naming scheme (tx_dlf_domain_model_document for tx_dlf_document). Maybe you could add a comment about this. We can discuss this before 4.0 release.

Reply:
I wasn't sure what the best place would be. For now, I put some remarks into a new file called "DEVELOPMENT.md".

Reply:
Unfortunately, this change has a side effect with TYPO3 9.

* go to the backend module and edit a document of type "year" (e.g. uid=15 if you are using the same data sets)
* save the record

This will cause the following error:

![New TYPO3 site  TYPO3 CMS 9 5 31  (1)](https://user-images.githubusercontent.com/3354500/149415972-a5782d67-eab6-492c-a708-fc467e52a1b5.png)

The "missing" table looks like a mapping fake. It's true, that this table does not exists.

This does not happen without your PR and TYPO3 9.
This does not happen with your PR and TYPO3 10.

Hope, you can reproduce this behavior.

Reply:
Very strange. I can't reproduce the behavior from last friday anymore :-(